api/src/main/java/bio/terra/tanagra/service/query/EntityCounts.java

```diff
@@ -0,0 +1,85 @@
+package bio.terra.tanagra.service.query;
+
+import bio.terra.tanagra.service.search.Attribute;
+import bio.terra.tanagra.service.search.EntityVariable;
+import bio.terra.tanagra.service.search.Filter;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A query for counts of entity instances. Contains the number of entity instances in each group,
+ * and a list of the attribute values that define each group.
+ *
+ * <p>e.g. {20, [Male, American]}, {35, [Female, Japanese]}
+ */
+@AutoValue
+public abstract class EntityCounts {
+  /** The primary entity and variable that the counts are being created for. */
+  public abstract EntityVariable primaryEntity();
+
+  /**
+   * Any additional attributes to return as columns of the dataset. These must be attributes of the
+   * primary entity.
+   */
+  @Nullable
+  public abstract ImmutableList<Attribute> additionalSelectedAttributes();
+
+  /** The attributes to group by. These must be attributes of the primary entity. */
+  @Nullable
+  public abstract ImmutableList<Attribute> groupByAttributes();
+
+  /** The filter to apply to the primary entity. */
+  public abstract Filter filter();
+
+  public static Builder builder() {
+    return new AutoValue_EntityCounts.Builder();
+  }
+
+  /** Builder for {@link EntityCounts}. */
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder primaryEntity(EntityVariable primaryEntity);
+
+    public abstract EntityVariable primaryEntity();
+
+    public abstract Builder additionalSelectedAttributes(List<Attribute> groupByAttributes);
+
+    public abstract ImmutableList<Attribute> additionalSelectedAttributes();
+
+    public abstract Builder groupByAttributes(List<Attribute> groupByAttributes);
```

Comment:
Are there restrictions around this, e.g. only for indexed counts? If so - where is this enforced?

Reply:
No restrictions, since this won't touch any of the indexed/precomputed rollup counts. These APIs are specifically for the dynamic rollup counts = how many men/women there are in my cohort. We can't precompute this because it depends on the definition of the cohort. By contrast, static count = how many people have a condition. We can precompute that because it doesn't depend on a cohort.

Reply:
OK, I ask this because it seems like I could pretty easily construct a query of death by accident or intentionally by doing something dumb like: GROUP BY condition_concept_id

I'm wondering whether there should be some kind of reasonably high limit or constraint on the number of buckets this generates. e.g. only returns up to the top 10K buckets, or LIMIT the group by query to 10K and fail if we reach said limit in results. Something to avoid OOMs here.

Reply:
Good point. Right now, I think we'll only ever get a single page size of results back, similar to what happens when the UI requests e.g. a list of all conditions, so I don't think OOMs will be an immediate problem. Longer-term, here are 2 things we can do that will address this:
1. Include a limit field for all endpoints that return query results. Set the default to something that won't cause OOMs e.g. 100.
2. Specify in the underlay which attributes we want to allow grouping by. The UI will need this somehow anyway, so it knows what attributes to show in the breakdown panel. Then when the UI hits the count API endpoint, we can validate that all of groupByAttributes are allowable. Presumably this will avoid queries of death because a human has indicated that grouping by this attribute is reasonable.